go/arrow/csv/reader.go

```diff
@@ -224,142 +246,185 @@ func (r *Reader) validate(recs []string) {
-		case *arrow.Float32Type:
-			v := r.readF32(str)
-			r.bld.Field(i).(*array.Float32Builder).Append(v)
-		case *arrow.Float64Type:
-			v := r.readF64(str)
-			r.bld.Field(i).(*array.Float64Builder).Append(v)
-		case *arrow.StringType:
-			r.bld.Field(i).(*array.StringBuilder).Append(str)
+func (r *Reader) isNull(val string) bool {
+	for _, ns := range r.nullValues {
+		if val == ns {
+			return true
 		}
 	}
+	return false
 }
 
-func (r *Reader) readI8(str string) int8 {
-	v, err := strconv.ParseInt(str, 10, 8)
-	if err != nil && r.err == nil {
-		r.err = err
-		return 0
+func (r *Reader) read(recs []string) {
+	for i, str := range recs {
+		r.fieldConverter[i](r.bld.Field(i), str)
 	}
-	return int8(v)
 }
 
-func (r *Reader) readI16(str string) int16 {
-	v, err := strconv.ParseInt(str, 10, 16)
-	if err != nil && r.err == nil {
-		r.err = err
-		return 0
+func (r *Reader) initFieldConverter(field *arrow.Field) func(array.Builder, string) {
+	switch field.Type.(type) {
+	case *arrow.BooleanType:
+		return func(field array.Builder, str string) {
+			r.parseBool(field, str)
+		}
+	case *arrow.Int8Type:
+		return func(field array.Builder, str string) {
+			v, null := r.parseInt(field, str, 8)
+			if !null {
+				field.(*array.Int8Builder).Append(int8(v))
+			}
+		}
+	case *arrow.Int16Type:
+		return func(field array.Builder, str string) {
+			v, null := r.parseInt(field, str, 16)
+			if !null {
+				field.(*array.Int16Builder).Append(int16(v))
+			}
+		}
+	case *arrow.Int32Type:
+		return func(field array.Builder, str string) {
+			v, null := r.parseInt(field, str, 32)
+			if !null {
+				field.(*array.Int32Builder).Append(int32(v))
+			}
+		}
+	case *arrow.Int64Type:
+		return func(field array.Builder, str string) {
+			v, null := r.parseInt(field, str, 64)
+			if !null {
+				field.(*array.Int64Builder).Append(int64(v))
+			}
+		}
+	case *arrow.Uint8Type:
+		return func(field array.Builder, str string) {
+			v, null := r.parseUint(field, str, 8)
+			if !null {
+				field.(*array.Uint8Builder).Append(uint8(v))
+			}
+		}
+	case *arrow.Uint16Type:
+		return func(field array.Builder, str string) {
+			v, null := r.parseUint(field, str, 16)
+			if !null {
+				field.(*array.Uint16Builder).Append(uint16(v))
+			}
+		}
+	case *arrow.Uint32Type:
+		return func(field array.Builder, str string) {
+			v, null := r.parseUint(field, str, 32)
+			if !null {
+				field.(*array.Uint32Builder).Append(uint32(v))
+			}
+		}
+	case *arrow.Uint64Type:
+		return func(field array.Builder, str string) {
+			v, null := r.parseUint(field, str, 64)
+			if !null {
+				field.(*array.Uint64Builder).Append(uint64(v))
+			}
+		}
+	case *arrow.Float32Type:
+		return func(field array.Builder, str string) {
+			v, null := r.parseFloat(field, str, 32)
+			if !null {
+				field.(*array.Float32Builder).Append(float32(v))
+			}
+		}
+	case *arrow.Float64Type:
+		return func(field array.Builder, str string) {
+			v, null := r.parseFloat(field, str, 64)
```

Comment:
ditto (`parseF32` and `parseF64`) ?

Reply:
I believe I did this with slightly longer names.